csharp/src/Ice/ConnectionI.cs

```diff
@@ -1447,6 +1450,39 @@ internal void sendHeartbeat()
         {
             if (isActiveOrHolding())
             {
+                // We check if the connection has become inactive.
+                if (
+                    _inactivityTimer is null &&           // timer not already scheduled
+                    _inactivityTimeout > TimeSpan.Zero && // inactivity timeout is enabled
+                    _state == StateActive &&              // only schedule the timer if the connection is active
+                    _dispatchCount == 0 &&                // no pending dispatch
+                    _asyncRequests.Count == 0 &&          // no pending invocation
+                    _readHeader)                          // we're not waiting for the remainder of an incoming message
+                {
+                    // We may become inactive while the peer is back-pressuring us. In this case, we only schedule the
+                    // inactivity timer if all outgoing messages in _sendStreams are heartbeats.
```

Comment:
While this might be theoretically possible it sounds like a very odd situation. Why would the peer back-pressure when there are only heartbeats?

Reply:
The peer doesn't need a reason to back-pressure us. It can do it at any time.

Reply:
Also, and that's very important, in this situation _sendStreams.Count will remain > 0 for a while (as long as we're back pressured). And this should not disable the inactivity timeout.